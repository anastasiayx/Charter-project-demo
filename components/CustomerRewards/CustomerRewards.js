import React from 'react';
import { arrayGroupByPropertyValue, sumArrayItemPropertyNumValue } from '../../utils/array.helper'
import { Collapse, Table } from 'antd'

const CustomerRewardsTableColumn = [

    {
        title: 'Transaction ID',
        dataIndex: 't_id',
        key: 't_id',
    },
    {
        title: 'Customer ID',
        dataIndex: 'c_id',
        key: 'c_id',
    },
    {
        title: "Spent Price",
        dataIndex: 'spent',
        key: 'spent',
    },
    {
        title: "Create Date",
        dataIndex: 'create_date',
        key: 'create_date',
    },
    {
        title: "Rewards Points",
        dataIndex: 'rewards',
        key: 'rewards',
    },

]


const CustomerRewards = ({ transactions }) => {
    const formatDateTransctions = transactions.map(transction => {
        return {
            ...transction,
            key: transction.t_id,
            formatDate: transction.create_date ? transction.create_date.substring(0, 7) : undefined
        }
    })
    const formatDateMap = arrayGroupByPropertyValue(formatDateTransctions, 'formatDate');

    return <Collapse defaultActiveKey="panel-month-0">
        {
            Object.entries(formatDateMap).map(([formatDate, formatDateTransactions], index) => {
                return <Collapse.Panel header={`${formatDate}, Monthly Rewards:${sumArrayItemPropertyNumValue(formatDateTransactions, 'rewards')}`} key={`panel-month-${index}`}>
                    <Table dataSource={formatDateTransactions} columns={CustomerRewardsTableColumn} />
                </Collapse.Panel>
            })
        }
    </Collapse>


}


export default React.memo(CustomerRewards);