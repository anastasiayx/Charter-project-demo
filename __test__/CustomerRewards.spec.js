import { render } from '@testing-library/react'
import CustomerRewards from '../components/CustomerRewards/CustomerRewards'
import './matchMediaMock'
describe('TransactionsRewards', () => {
    const customerTransactionsMock = [
        {
            t_id: "1",
            spent: 130,
            create_date: new Date("8/29/22").toISOString(),
            rewards: 110,
            c_id: "sdfs34234fss"
        },
        {
            t_id: "2",
            spent: 200,
            create_date: new Date("7/29/22").toISOString(),
            rewards: 250,
            c_id: "sdfs34234fss"
        },
        {
            t_id: "3",
            spent: 30,
            create_date: new Date("6/29/22").toISOString(),
            rewards: 0,
            c_id: "sdfs34234fss"
        },
        {
            t_id: "4",
            spent: 187,
            create_date: new Date("7/29/22").toISOString(),
            rewards: 224,
            c_id: "sdfs34234fss"
        },
        {
            t_id: "9",
            spent: 187,
            create_date: new Date("8/29/22").toISOString(),
            rewards: 224,
            c_id: "sdfs34234fss"
        }
    ]

    it('should match snap shot', () => {
        const { asFragment } = render(<CustomerRewards transactions={customerTransactionsMock} />)
        expect(asFragment()).toMatchSnapshot()
    })

    it('should render customer pannels group by yyyy/mm/dd', () => {
        const { getByText } = render(<CustomerRewards transactions={customerTransactionsMock} />)
        expect(getByText("2022-08, Monthly Rewards:334")).toBeInTheDocument()
        expect(getByText("2022-07, Monthly Rewards:474")).toBeInTheDocument()
        expect(getByText("2022-06, Monthly Rewards:0")).toBeInTheDocument()
    })
})