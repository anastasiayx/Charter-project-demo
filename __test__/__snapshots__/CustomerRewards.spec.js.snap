// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionsRewards should match snap shot 1`] = `
<DocumentFragment>
  <div
    class="ant-collapse ant-collapse-icon-position-start"
  >
    <div
      class="ant-collapse-item ant-collapse-item-active"
    >
      <div
        aria-disabled="false"
        aria-expanded="true"
        class="ant-collapse-header"
        role="button"
        tabindex="0"
      >
        <div
          class="ant-collapse-expand-icon"
        >
          <span
            aria-label="right"
            class="anticon anticon-right ant-collapse-arrow"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="right"
              fill="currentColor"
              focusable="false"
              height="1em"
              style="transform: rotate(90deg);"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
              />
            </svg>
          </span>
        </div>
        <span
          class="ant-collapse-header-text"
        >
          2022-08, Monthly Rewards:334
        </span>
      </div>
      <div
        class="ant-collapse-content ant-collapse-content-active"
      >
        <div
          class="ant-collapse-content-box"
        >
          <div
            class="ant-table-wrapper"
          >
            <div
              class="ant-spin-nested-loading"
            >
              <div
                class="ant-spin-container"
              >
                <div
                  class="ant-table"
                >
                  <div
                    class="ant-table-container"
                  >
                    <div
                      class="ant-table-content"
                    >
                      <table
                        style="table-layout: auto;"
                      >
                        <colgroup />
                        <thead
                          class="ant-table-thead"
                        >
                          <tr>
                            <th
                              class="ant-table-cell"
                            >
                              Transaction ID
                            </th>
                            <th
                              class="ant-table-cell"
                            >
                              Customer ID
                            </th>
                            <th
                              class="ant-table-cell"
                            >
                              Spent Price
                            </th>
                            <th
                              class="ant-table-cell"
                            >
                              Create Date
                            </th>
                            <th
                              class="ant-table-cell"
                            >
                              Rewards Points
                            </th>
                          </tr>
                        </thead>
                        <tbody
                          class="ant-table-tbody"
                        >
                          <tr
                            class="ant-table-row ant-table-row-level-0"
                            data-row-key="1"
                          >
                            <td
                              class="ant-table-cell"
                            >
                              1
                            </td>
                            <td
                              class="ant-table-cell"
                            >
                              sdfs34234fss
                            </td>
                            <td
                              class="ant-table-cell"
                            >
                              130
                            </td>
                            <td
                              class="ant-table-cell"
                            >
                              2022-08-29T04:00:00.000Z
                            </td>
                            <td
                              class="ant-table-cell"
                            >
                              110
                            </td>
                          </tr>
                          <tr
                            class="ant-table-row ant-table-row-level-0"
                            data-row-key="9"
                          >
                            <td
                              class="ant-table-cell"
                            >
                              9
                            </td>
                            <td
                              class="ant-table-cell"
                            >
                              sdfs34234fss
                            </td>
                            <td
                              class="ant-table-cell"
                            >
                              187
                            </td>
                            <td
                              class="ant-table-cell"
                            >
                              2022-08-29T04:00:00.000Z
                            </td>
                            <td
                              class="ant-table-cell"
                            >
                              224
                            </td>
                          </tr>
                        </tbody>
                      </table>
                    </div>
                  </div>
                </div>
                <ul
                  class="ant-pagination ant-table-pagination ant-table-pagination-right"
                  unselectable="unselectable"
                >
                  <li
                    aria-disabled="true"
                    class="ant-pagination-prev ant-pagination-disabled"
                    title="Previous Page"
                  >
                    <button
                      class="ant-pagination-item-link"
                      disabled=""
                      tabindex="-1"
                      type="button"
                    >
                      <span
                        aria-label="left"
                        class="anticon anticon-left"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="left"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                          />
                        </svg>
                      </span>
                    </button>
                  </li>
                  <li
                    class="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
                    tabindex="0"
                    title="1"
                  >
                    <a
                      rel="nofollow"
                    >
                      1
                    </a>
                  </li>
                  <li
                    aria-disabled="true"
                    class="ant-pagination-next ant-pagination-disabled"
                    title="Next Page"
                  >
                    <button
                      class="ant-pagination-item-link"
                      disabled=""
                      tabindex="-1"
                      type="button"
                    >
                      <span
                        aria-label="right"
                        class="anticon anticon-right"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="right"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                          />
                        </svg>
                      </span>
                    </button>
                  </li>
                </ul>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-collapse-item"
    >
      <div
        aria-disabled="false"
        aria-expanded="false"
        class="ant-collapse-header"
        role="button"
        tabindex="0"
      >
        <div
          class="ant-collapse-expand-icon"
        >
          <span
            aria-label="right"
            class="anticon anticon-right ant-collapse-arrow"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="right"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
              />
            </svg>
          </span>
        </div>
        <span
          class="ant-collapse-header-text"
        >
          2022-07, Monthly Rewards:474
        </span>
      </div>
    </div>
    <div
      class="ant-collapse-item"
    >
      <div
        aria-disabled="false"
        aria-expanded="false"
        class="ant-collapse-header"
        role="button"
        tabindex="0"
      >
        <div
          class="ant-collapse-expand-icon"
        >
          <span
            aria-label="right"
            class="anticon anticon-right ant-collapse-arrow"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="right"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
              />
            </svg>
          </span>
        </div>
        <span
          class="ant-collapse-header-text"
        >
          2022-06, Monthly Rewards:0
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
